packages/calcite-components/src/utils/focusTrapComponent.ts

```diff
@@ -58,9 +71,10 @@ export function connectFocusTrap(component: FocusTrapComponent, options?: Connec
   }
 
   const focusTrapOptions: FocusTrapOptions = {
-    clickOutsideDeactivates: true,
-    escapeDeactivates: false,
+    clickOutsideDeactivates: component.clickOutsideDeactivates ?? true,
+    escapeDeactivates: !component.escapeDisabled ?? true,
```

Comment:
FIXME: I noticed one issue with us relying on `focus-trap`'s `escapeDeactivates` feature. Since it uses the [`capture` phase for `keydown` events](https://github.com/focus-trap/focus-trap/blob/master/index.js#L837), it means our components can no longer ignore `Escape` presses that were already handled by a child component (e.g., user presses `Escape` to close open combobox inside a modal).

Reply:
hmm that's not good. 

It seems to have been that way from the beginning (first release)

We could ask and see why its that way and if there is any chance of having a focusTrap option to not use capture events.

Reply:
related issue: https://github.com/focus-trap/focus-trap/issues/1165

Reply:
I guess instead of using escapeDeactivates, we would have to handle closing the appropriate focus trap element in the trapStack

Would have to look at our trapStack and call trap.deactivate()
 

https://github.com/focus-trap/focus-trap?tab=readme-ov-file#trapdeactivate

This means we would set `escapeDeactivates` to false and each component would handle calling some global method for closing the appropriate focusTrap on the trapStack.

Reply:
We would need to setup an event listener on the document to listen for the escape key and close the top most focusTrap in the focusTrapStack. We could probably have a method in the `focusTrapComponent` to do this.

To setup the event listener, we'll need to track how many active focus traps there are by having some trap count in `connectFocusTrap`. We'll need to add a method `disconnectFocusTrap` to be able to remove the event listener if there are no more focus traps in the stack.

Reply:
Here's an example of what I'm thinking:

```patch
diff --git forkSrcPrefix/packages/calcite-components/src/utils/focusTrapComponent.ts forkDstPrefix/packages/calcite-components/src/utils/focusTrapComponent.ts
index 877c3dc04f3329dd1355a30e8af3d888b3e96cf4..e88af0d7cdf6c530382888b3f8a02c5537bc6dd7 100644
--- forkSrcPrefix/packages/calcite-components/src/utils/focusTrapComponent.ts
+++ forkDstPrefix/packages/calcite-components/src/utils/focusTrapComponent.ts
@@ -2,6 +2,9 @@ import { createFocusTrap, FocusTrap as _FocusTrap, Options as FocusTrapOptions }
 import { FocusableElement, focusElement, tabbableOptions } from "./dom";
 import { focusTrapStack } from "./config";
 
+// i'm not sure if this needs to be a map of all documents. Would the keydown event bubble up out of shadow dom elements? If so, then the map isn't necessary.
+const focusTrapDocumentMap = new Map<Document, number>();
+
 /**
  * Defines interface for components with a focus trap. Focusable content is required for components implementing focus trapping with this interface.
  */
@@ -51,6 +54,7 @@ interface ConnectFocusTrapOptions {
  */
 export function connectFocusTrap(component: FocusTrapComponent, options?: ConnectFocusTrapOptions): void {
   const { el } = component;
+  const { ownerDocument } = el;
   const focusTrapNode = options?.focusTrapEl || el;
 
   if (!focusTrapNode) {
@@ -68,12 +72,28 @@ export function connectFocusTrap(component: FocusTrapComponent, options?: Connec
     ...options?.focusTrapOptions,
 
     // the following options are not overridable
-    document: el.ownerDocument,
+    document: ownerDocument,
     tabbableOptions,
     trapStack: focusTrapStack,
   };
 
   component.focusTrap = createFocusTrap(focusTrapNode, focusTrapOptions);
+  const newCount = (focusTrapDocumentMap.get(ownerDocument) || 0) + 1;
+  if (newCount === 1) {
+    addEscapeListener(ownerDocument);
+  }
+  focusTrapDocumentMap.set(ownerDocument, newCount);
+}
+
+// would need to call this on disconnect of all focus trap components.
+export function disconnectFocusTrap(component: FocusTrapComponent): void {
+  const { el } = component;
+  const { ownerDocument } = el;
+  const newCount = (focusTrapDocumentMap.get(ownerDocument) || 1) - 1;
+  if (newCount === 0) {
+    removeEscapeListener(ownerDocument);
+  }
+  focusTrapDocumentMap.set(ownerDocument, newCount);
 }
 
 /**
@@ -119,3 +139,17 @@ export function deactivateFocusTrap(
 export function updateFocusTrapElements(component: FocusTrapComponent): void {
   component.focusTrap?.updateContainerElements(component.el);
 }
+
+function escapeHandler(event: KeyboardEvent): void {
+  if (event.key === "Escape") {
+    focusTrapStack[focusTrapStack.length - 1]?.deactivate(); // deactivate the last active focus trap. This might be the 0 index. I'm not sure.
+  }
+}
+
+function addEscapeListener(document: Document): void {
+  document.addEventListener("keydown", escapeHandler);
+}
+
+function removeEscapeListener(document: Document): void {
+  document.removeEventListener("keydown", escapeHandler);
+}
```

This would let us handle the escape key and close the last active trap.

Reply:
Example: https://github.com/Esri/calcite-design-system/pull/10197